arangosh/Import/AutoTuneThread.cpp

```diff
@@ -0,0 +1,138 @@
+///
+/// @author Matthew Von-Maszewski
+////////////////////////////////////////////////////////////////////////////////
+
+#include <iostream>
+#include <thread>
+
+#include "AutoTuneThread.h"
+#include "Basics/ConditionLocker.h"
+#include "ImportHelper.h"
+
+using namespace arangodb;
+using namespace arangodb::import;
+
+////////////////////////////////////////////////////////////////////////////////
+// Goals:
+//  1. compute current one second throughput of import
+//  2. spread byte count of one second throughput across sender threads
+//  3. create "space" between sender execution to give server time for other activities
+//
+// The code collects the total count of bytes absorbed for ten seconds, then averages
+//  that amount with the total from the previous 10 seconds.  The per second per thread pace
+//  is therefore average divided by the thread count divided by 10.
+//
+// The pace starts "slow", 1 megabyte per second.  Each recalculation of pace adds
+//  a 20% growth factor above the actual calculation from average bytes consumed.
+//
+// The pacing code also notices when threads are completing quickly.  It will release
+//  a new thread early in such cases to again encourage rate growth.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+AutoTuneThread::AutoTuneThread(ImportHelper & importHelper)
+    : Thread("AutoTuneThread"),
+      _importHelper(importHelper),
+      _nextSend(std::chrono::steady_clock::now()),
+      _pace(std::chrono::microseconds(1000000 / importHelper.getThreadCount()))
+{}
+
+AutoTuneThread::~AutoTuneThread() {
+  shutdown();
+}
+
+void AutoTuneThread::beginShutdown() {
+  Thread::beginShutdown();
+
+  // wake up the thread that may be waiting in run()
+  CONDITION_LOCKER(guard, _condition);
+  guard.broadcast();
+}
+
+
+void AutoTuneThread::run() {
+  while (!isStopping()) {
+    {
+      CONDITION_LOCKER(guard, _condition);
+      guard.wait(std::chrono::seconds(10));
+    }
+    if (!isStopping()) {
+      // getMaxUploadSize() is per thread
+      uint64_t current_max = _importHelper.getMaxUploadSize();
+      current_max *= _importHelper.getThreadCount();
+      uint64_t ten_second_actual = _importHelper.rotatePeriodByteCount();
+      uint64_t new_max = current_max;
+
+      // is current_max way too big
+      if (ten_second_actual < current_max && 10 < ten_second_actual) {
+        new_max = ten_second_actual / 10;
+      } else if ( ten_second_actual <= 10 ) {
+        new_max = current_max / 10;
+      } else {
+        new_max = (current_max + ten_second_actual / 10) / 2;
+      }
+
+      // grow number slowly if possible (10%)
+//      new_max += new_max/10;
+      new_max += new_max/5;  //(20% growth)
+
+      // make "per thread"
+      new_max /= _importHelper.getThreadCount();
+
+      // notes in Import mention an internal limit of 768MBytes
+      if ((768 * 1024 * 1024) < new_max) {
+        new_max = 768 * 1024 * 1024;
```

Comment:
Is the 768MB here the same as ImportHelper::MaxBatchSize? Then we should refer to this constant here instead of hard-coding a magic number.

Reply:
Made more changes that suspected.  Was ImportFeature::MaxBatchSize.  Had to create the constant within ImportHelper so that both ImportFeature and AutoTuneThread could find it in both ways import code gets linked.